From d871eb23d46c08fbf4c2780fa3ebd1c0ef872cd7 Mon Sep 17 00:00:00 2001
From: Jehan <jehan@girinstud.io>
Date: Mon, 14 Feb 2022 14:13:57 +0100
Subject: [PATCH] gdk: do not draw contents outside the rect passed by
 drawRect().

Apple docs says:

> The dirtyRect parameter helps you achieve better performance by
> specifying the portion of the view that needs to be drawn. You should
> always limit drawing to the content inside this rectangle.

We were not doing this, basically limiting our testing with rect to
whether it is empty. We should actually combine it with values returned
by getRectsBeingDrawn, and only draw intersections.
---
 gdk/quartz/GdkQuartzView.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/gdk/quartz/GdkQuartzView.c b/gdk/quartz/GdkQuartzView.c
index ca68e031ac..dc15aaf286 100644
--- a/gdk/quartz/GdkQuartzView.c
+++ b/gdk/quartz/GdkQuartzView.c
@@ -318,7 +318,7 @@
   if (! (gdk_window->event_mask & GDK_EXPOSURE_MASK))
     return;
 
-  if (NSEqualRects (rect, NSZeroRect))
+  if (NSIsEmptyRect (rect))
     return;
 
   if (!GDK_WINDOW_IS_MAPPED (gdk_window))
@@ -353,11 +353,19 @@
     {
       cairo_region_t        *region;
       cairo_rectangle_int_t  cairo_rect;
+      NSRect                 intersect_rect;
 
-      cairo_rect.x = drawn_rects[i].origin.x;
-      cairo_rect.y = drawn_rects[i].origin.y;
-      cairo_rect.width = drawn_rects[i].size.width;
-      cairo_rect.height = drawn_rects[i].size.height;
+      /* Apple docs of drawRect() clearly says we should always limit drawing
+       * to contents inside the rectangle passed as parameter.
+       */
+      intersect_rect = NSIntersectionRect (rect, drawn_rects[i]);
+      if (NSIsEmptyRect (intersect_rect))
+        continue;
+
+      cairo_rect.x = intersect_rect.origin.x;
+      cairo_rect.y = intersect_rect.origin.y;
+      cairo_rect.width = intersect_rect.size.width;
+      cairo_rect.height = intersect_rect.size.height;
 
       region = cairo_region_create_rectangles (&cairo_rect, 1);
 
-- 
2.30.2

